tensorboard/components/tf_categorization_utils/categorizationUtils.ts

```diff
@@ -110,25 +122,102 @@ export function categorizeTags(
     query?: string): TagCategory[] {
   const tags = tf_backend.getTags(runToTag);
   const categories = categorize(tags, query);
-  const tagToRuns: {[tag: string]: string[]} = {};
-  tags.forEach(tag => {
-    tagToRuns[tag] = [];
-  });
-  selectedRuns.forEach(run => {
-    (runToTag[run] || []).forEach(tag => {
-      tagToRuns[tag].push(run);
-    });
-  });
+  const tagToRuns = createTagToRuns(_.pick(runToTag, selectedRuns));
+
   return categories.map(({name, metadata, items}) => ({
     name,
     metadata,
     items: items.map(tag => ({
       tag,
-      runs: tagToRuns[tag].slice(),
+      runs: tagToRuns.get(tag).slice(),
     })),
   }));
 }
 
+/**
+ * Creates grouping of the data based on selection from tf-data-selector. It
+ * groups data by prefixes of tag names and by tag names. Each group contains
+ * series, a tuple of experiment name and run name.
+ */
+export function categorizeSelection(
+    selection: tf_data_selector.Selection[], pluginName: string):
+    SeriesCategory[] {
+  const tagToSeries = new Map();
+  const searchTags = new Set();
```

Comment:
I think we probably want to resort searchTags at some point so that the result is still in "tag-comparison order" (same as now) within the "search" category pane.  Right now if I'm not mistaken, each individual batch of tags is sorted but they aren't sorted overall, and Set just preserves insertion order, so you might get a result like "aa, zz, ab, ac" if "aa, zz" are in the first experiment and "aa, ab, ac" are in the second one, which seems undesirable.

Maybe good to have a test for this case too.

Reply:
Done! Thanks for the insight